spec/fixtures/mocks/lua-resty-dns/resty/dns/resolver.lua

```diff
@@ -63,43 +63,51 @@ do
 
     local filename = prefix .. "/" .. MOCK_RECORD_FILENAME
 
-    mock_records = {}
+    mock_file = {}
     if not is_file(filename) then
       -- no mock records set up, return empty default
       ngx.log(ngx.DEBUG, LOG_PREFIX, "bypassing mock, no mock records found")
-      return mock_records
+      return mock_file
     end
 
     -- there is a file with mock records available, go load it
     local f = assert(io.open(filename))
     local json_file = assert(f:read("*a"))
     f:close()
 
-    mock_records = assert(cjson.decode(json_file))
-
-    return mock_records
+    mock_file = assert(cjson.decode(json_file))
+    return mock_file.records, mock_file.should_fail
   end
 end
 
 
 -- patch the actual query method
 local old_query = resolver.query
 resolver.query = function(self, name, options, tries)
-  local mock_records = get_mock_records()
+  local mock_records, should_fail = get_mock_records()
   local qtype = (options or {}).qtype or resolver.TYPE_A
 
   local answer = (mock_records[qtype] or {})[name]
   if answer then
     -- we actually have a mock answer, return it
     ngx.log(ngx.DEBUG, LOG_PREFIX, "serving '", name, "' from mocks")
-    ngx.timer.at(0.05, function() ngx.log(ngx.INFO, "tick") end) -- FIXME: remove after OpenResty semaphore bug is fixed
     return answer, nil, tries
   end
 
-  -- no mock, so invoke original resolver
-  local a, b, c = old_query(self, name, options, tries)
-  ngx.timer.at(0.05, function() ngx.log(ngx.INFO, "tick") end) -- FIXME: remove after OpenResty semaphore bug is fixed
-  return a, b, c
+  if not should_fail then
+    -- no mock, so invoke original resolver
+    local a, b, c = old_query(self, name, options, tries)
+    return a, b, c
+  end
+end
+
+do
+  local semaphore = require "ngx.semaphore"
```

Comment:
this looks as something that should go in `globalpatches.lua`, any specific reasons not to put it there @javierguerragiraldez  ?

Reply:
so far, we've only seen this happen in the CI test, so I wanted to constrain the change as much as possible.  I'm not sure the openresty bug will be accepted as a bug, so instead of adding yet another patch, we might have to add the `sleep(0)` to the dns client module.   or make a rule of thumb like "if you `:post()` from a timer, be sure to yield before returning"

similar thing with the other part of this fix: i'm pretty sure a mock DNS shouldn't fall back to the real one, but that is the current behaviour, so i made it conditional (and default) to avoid unintended effects

Reply:
>  I'm not sure the openresty bug will be accepted as a bug

are we planning on providing a patch/PR for upstream?

can we at least add a TODO or FIXME comment here to easily track this tech debt?

Reply:
> similar thing with the other part of this fix: i'm pretty sure a mock DNS shouldn't fall back to the real one, but that is the current behaviour, so i made it conditional (and default) to avoid unintended effects

It was created to create not as mock-DNS, but as mock-DNS-records, but I think it makes a nice addition. The only thing missing is the doc-comments describing the additional option (in `helpers.lua`). To check the docs do;
```
cd spec
ldoc .
open docs/index.html
```
So what I get is that the new-default is to NOT fall-back, which is breaking behaviour (from test helper perspective) right? This probably means we should also version the test-helpers.

Reply:
no, the default is to behave as previously.  only this one test prevents the mock from going into the real world.